src/org/mozilla/javascript/Hashtable.java

```diff
@@ -48,9 +49,16 @@
         }
 
         Entry(Object k, Object value) {
-            if ((k instanceof Number) && (!(k instanceof Double))) {
-                // Hash comparison won't work if we don't do this
-                this.key = Double.valueOf(((Number) k).doubleValue());
+            if (k instanceof Number) {
+                if (k instanceof Double || k instanceof BigInteger) {
+                    // BigInteger needs to retain its own type, due to
+                    // "If Type(x) is different from Type(y), return false." in
+                    // ecma262/multipage/abstract-operations.html#sec-samevaluezero
+                    this.key = k;
```

Comment:
Sorry but maybe one minor point. This method uses sometimes the this.key notation and sometimes only key.
I think we can cleanup the code here a bit and always use key without 'this.'

Many thanks for the fix and the test cases.